packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -49,7 +55,95 @@ export function synthesize(root: IConstruct, options: SynthesisOptions = { }): c
   // stacks to add themselves to the synthesized cloud assembly.
   synthesizeTree(root, builder, options.validateOnSynthesis);
 
-  return builder.buildAssembly();
+  const assembly = builder.buildAssembly();
+
+  invokeValidationPlugins(root, builder.outdir);
+
+  return assembly;
+}
+
+class LazyHash {
+  private hash?: string;
+
+  constructor(private readonly outdir: string) {}
+
+  get value(): string {
+    if (!this.hash) {
+      this.hash = computeChecksumOfFolder(this.outdir);
+    }
+    return this.hash;
+  }
+}
+
+function invokeValidationPlugins(root: IConstruct, outdir: string) {
+  const lazyHash = new LazyHash(outdir);
+  let failed = false;
+
+  const templatePathsByPlugin: Map<IValidationPlugin, string[]> = new Map();
+  visit(root, 'post', construct => {
+    if (Stage.isStage(construct)) {
+      for (const plugin of construct.validationPlugins) {
+        if (!templatePathsByPlugin.has(plugin)) {
+          templatePathsByPlugin.set(plugin, []);
+        }
+        templatePathsByPlugin.get(plugin)!.push(...templatePaths(construct.outdir));
+      }
+    }
+  });
+
+  const reports: ValidationReport[] = [];
+  for (const [plugin, paths] of templatePathsByPlugin.entries()) {
+    const originalHash = lazyHash.value;
+
+    if (!plugin.isReady()) {
+      throw new Error(`Validation plugin '${plugin.name}' is not ready`);
```

Comment:
Instead of failing here, just add this to the report.

Reply:
Done.